src/compiler/checker.ts

```diff
@@ -47308,9 +47316,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // Grammar checking
             checkGrammarSourceFile(file);
 
+            clear(potentialThisCollisions);

```

Comment:
Honestly I think this whole "one giant array kept until the end" is super weird and I've been wondering if we could refactor this to not use this pattern...

Reply:
That could make my life more convenient for this PR.